### PR TITLE
feat: Add support for defining width and height for friendly healthbars

### PR DESCRIPTION
**Pull request by kristijanhusak**
This addresses some of the following issues:

* https://github.com/Backupiseasy/ThreatPlates/issues/158 (partially)
* https://github.com/Backupiseasy/ThreatPlates/issues/366

Disclaimer: This is my first time contributing to WoW addon, so please let me know if I missed something

### CODE REVIEW

**Comment by Backupiseasy** [comment]
It feels great to have someone contribute to Threat Plates. Thanks a lot! 

I will look into the pull request in the next days, but there are already two thinks that you can do now:

- The next relase with new features is 11.2.0. Can you please change the target branch to https://github.com/Backupiseasy/ThreatPlates/tree/devel/112/main. 
- It should not break the clickable area in retail as there are seperate clickable areas for friendly and enemy namepaltes. But in Classic, this is not the case so it might not work there (as you would not be able to click these nameplates reliable or as expected).

**Comment by kristijanhusak** [comment]
> It should not break the clickable area in retail as there are seperate clickable areas for friendly and enemy namepaltes. But in Classic, this is not the case so it might not work there (as you would not be able to click these nameplates reliable or as expected).

Can you give me a hint where and how should I handle this?

**Comment by Backupiseasy** [comment]
Generally speaking, the clickable area is set in SetBaseNamePlateSize (there a different function for Retail and Classic) in Core.lua. But I think it makes sense to first verify that it actually breaks in Classic. Just test in Classic with different nameplate sizes if the clickable area is still correct (you can check that by moving the mouse in proximity of the nameplate; once it's hightlighted, you are in the clickable area). 

Hm, looking at the code, I might be wrong in what I said above. It just seems to be an issue if Blizzard nameplates are enabled, but otherwise it should be fine, I guess.

**Comment by kristijanhusak** [comment]
That's actually an issue for both classic and retail. I'll take a look, thanks for the guidelines.

**Comment by kristijanhusak** [comment]
@Backupiseasy I fixed the clickable area for retail, but for classic it's not possible to have different clickareas depending on unit type, as you also mentioned in this comment https://github.com/Backupiseasy/ThreatPlates/blob/master/Core.lua#L70

Only suggestion that I have is to add a note to the Blizzard setting around healthbar sync explaining the issue.

**Comment by kristijanhusak** [comment]
One thing I noticed randomly happening in solo shuffle is that I get an enemy size for a friendly player, right after I had a match against them. Reload fixes it, so I'm assuming it's a cache issue. Do you have an idea how can I make sure it does not happen?

**Comment by Backupiseasy** [comment]
Ah, solo shuffle again. The main reason for that is that units switch from friendly to enemy and vice version from match to match. I had some issues with that, e.g., in the arena widget. What I did there is listening to ARENA_OPPONENT_UPDATE which is an event fired when this happens. Also UNIT_FACTION should be fired for all units I think. Maybe adding some code there could help (ARENA_OPPONENT_UPDATE is not yet processed in TidyPlatesCore.lua, so you might have to add it).

**Comment by kristijanhusak** [comment]
I added the ARENA_OPPONENT_UPDATE to just reset the nameplate size. Let me know if there's a better way to handle it. Didn't had issues so far.

**Comment by kristijanhusak** [comment]
> I added the ARENA_OPPONENT_UPDATE to just reset the nameplate size. Let me know if there's a better way to handle it. Didn't had issues so far.

Seems like this is not enough. It still happens from time to time to have the enemy size in solo shuffle.

**Comment by Backupiseasy** [comment]
I want to release a first beta for 10.2 this weekend. After that, I can look into this pull request. Sorry that it takes so long.

**Comment by kristijanhusak** [comment]
No worries. I'm using this anyway so there's no rush for me. Just reporting issues so you are aware of them once you get to it.

File: Options.lua [hunk 1]
```diff
@@ -6509,6 +6509,35 @@ local function CreateHealthbarOptions()
...
             }
           },
+          FriendlyUnitsFormat = {
+            name = L["Friendly Units"],
+            order = 30,
+            type = "group",
+            inline = true,
+            set = SetThemeValue,
+            args = {
+              Width = GetRangeEntry(L["Bar Width"], 10, { "settings", "healthbar", "widthFriendly" }, 5, 500,
```
Comment (Backupiseasy) [inline Options.lua:6519]: There is no default value for widthFriendly defined in Constants.lua

File: Options.lua [hunk 2]
```diff
@@ -6509,6 +6509,35 @@ local function CreateHealthbarOptions()
...
+                function(info, val)
+                  if InCombatLockdown() then
+                    Addon.Logging.Error(L["We're unable to change this while in combat"])
+                  else
+                    SetThemeValue(info, val)
+                    Addon:SetBaseNamePlateSize()
+                  end
+                end),
+              Height = GetRangeEntry(L["Bar Height"], 20, {"settings", "healthbar", "heightFriendly" }, 1, 100,
```
Comment (Backupiseasy) [inline Options.lua:6528]: There is no default value for heightFriendly defined in Constants.lua

File: Core.lua [hunk 3]
```diff
@@ -39,11 +39,18 @@ local LSMUpdateTimer
...
-    local height = db.healthbar.height
+    local width, height
+    if friendlyUnit then
+      width = db.healthbar.widthFriendly
+      height = db.healthbar.heightFriendly
+    else
+      width = db.healthbar.width
+      height = db.healthbar.height
+    end
+
```
Comment (Backupiseasy) [inline Core.lua:53]: I think I changed this code in a wrong way. In Classic, it's not possible to have separate nameplates sizes for friendly and enemy plates as the clickable area is always the same for both. This should be part of the non-classic code. I need to fix this afterwards.

File: Styles/basicThemes.lua [hunk 4]
```diff
@@ -21,6 +21,8 @@ local function Create(name)
       backdrop = Addon.LibSharedMedia:Fetch('statusbar', db.healthbar.backdrop, true),
       width = db.healthbar.width,
       height = db.healthbar.height,
+      widthFriendly = db.healthbar.widthFriendly or db.healthbar.width,
+      heightFriendly = db.healthbar.heightFriendly or db.healthbar.height,
```
Comment (Backupiseasy) [inline Styles/basicThemes.lua:25]: I think with default values defined, this "or" should not be possible.

File: TidyPlatesInternal/TidyPlatesCore.lua [hunk 5]
```diff
@@ -1327,6 +1327,10 @@ function CoreEvents:PLAYER_LOGIN()
   end
 end
 
+function CoreEvents:ARENA_OPPONENT_UPDATE()
+  Addon:SetBaseNamePlateSize()
+end
+
```
Comment (Backupiseasy) [inline TidyPlatesInternal/TidyPlatesCore.lua:1333]: Why is this here? Not sure how often ARENA_OPPONENT_UPDATE fires, but - if I don't remeber wrongly - several times, but that might be inefficient. Hopefully, there is a batter way.

File: TidyPlatesInternal/TidyPlatesCore.lua [hunk 6]
```diff
@@ -2084,7 +2088,18 @@ do
...
+      y = style.healthbar.y,
+      width = style.healthbar.width,
+      height = style.healthbar.height
+    }
+
+    if extended.unit.reaction == 'FRIENDLY' then
+      healthbarOpts.width = style.healthbar.widthFriendly
+      healthbarOpts.height = style.healthbar.heightFriendly
+    end
+    SetAnchorGroupObject(visual.healthbar, healthbarOpts, extended)
```
Comment (Backupiseasy) [inline TidyPlatesInternal/TidyPlatesCore.lua:2102]: I think it would be enought to change the style parameter (healthbarOpts) only if extended.unit.reaction == 'FRIENDLY', so basically only create the array in the if-part. Not sure though. Need to think about this a bit more.

File: TidyPlatesInternal/TidyPlatesCore.lua [hunk 7]
```diff
@@ -2229,9 +2244,16 @@ function Addon:ConfigClickableArea(toggle_show)
...
+    local background = extended.Background
+    local width, height
+    if extended.unit.reaction == "FRIENDLY" then
+      width, height = C_NamePlate.GetNamePlateFriendlySize()
+    else
+      width, height = C_NamePlate.GetNamePlateEnemySize()
+    end
     background:SetPoint("CENTER", ConfigModePlate.UnitFrame, "CENTER")
-    background:SetSize(Addon.db.profile.settings.frame.width, Addon.db.profile.settings.frame.height)
+    background:SetSize(width, height)
```
Comment (Backupiseasy) [inline TidyPlatesInternal/TidyPlatesCore.lua:2256]: I don't like it that there is still settings.frame, but now it only has the enemy width/height. Maybe friendlyWidth/friendlyHeight also needs to be stored there ...

**Review by Backupiseasy — COMMENTED** [review]
There are some things that I noticed, see my comments. I will still merge this change, I think, and update it afterwards (in branch feature/pr-459).

**Comment by kristijanhusak** [comment]
Hey, just a heads up that I stopped playing the game some time ago, so I'm not really able to contribute to this PR any more.

**Comment by Backupiseasy** [comment]
Understood. Still, thanks for contributing. I will still see how to best integrate this even if it's (very) late.
